File: backend/src/main/java/com/sweetshop/service/SweetService.java
package com.sweetshop.service;

import com.sweetshop.dto.SweetRequest;
import com.sweetshop.entity.Sweet;
import com.sweetshop.exception.ResourceNotFoundException;
import com.sweetshop.repository.SweetRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.List;

/**
 * Service for sweet-related operations.
 * Handles sweet management, inventory operations, and searches.
 */
@Service
@RequiredArgsConstructor
public class SweetService {

    private final SweetRepository sweetRepository;

    /**
     * Add a new sweet to the inventory.
     *
     * @param request the sweet request containing details
     * @return the created sweet
     */
    public Sweet addSweet(SweetRequest request) {
        Sweet sweet = Sweet.builder()
                .name(request.getName())
                .category(request.getCategory())
                .price(request.getPrice())
                .quantity(request.getQuantity())
                .description(request.getDescription())
                .build();

        return sweetRepository.save(sweet);
    }

    /**
     * Get all sweets from the inventory.
     *
     * @return list of all sweets
     */
    public List<Sweet> getAllSweets() {
        return sweetRepository.findAll();
    }

    /**
     * Get a sweet by its ID.
     *
     * @param id the sweet ID
     * @return the sweet
     * @throws ResourceNotFoundException if sweet is not found
     */
    public Sweet getSweetById(Long id) {
        return sweetRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Sweet not found with id: " + id));
    }

    /**
     * Update a sweet's details.
     *
     * @param id the sweet ID
     * @param request the updated sweet request
     * @return the updated sweet
     * @throws ResourceNotFoundException if sweet is not found
     */
    public Sweet updateSweet(Long id, SweetRequest request) {
        Sweet sweet = getSweetById(id);

        sweet.setName(request.getName());
        sweet.setCategory(request.getCategory());
        sweet.setPrice(request.getPrice());
        sweet.setQuantity(request.getQuantity());
        sweet.setDescription(request.getDescription());

        return sweetRepository.save(sweet);
    }

    /**
     * Delete a sweet from the inventory.
     *
     * @param id the sweet ID
     * @throws ResourceNotFoundException if sweet is not found
     */
    public void deleteSweet(Long id) {
        if (!sweetRepository.existsById(id)) {
            throw new ResourceNotFoundException("Sweet not found with id: " + id);
        }
        sweetRepository.deleteById(id);
    }

    /**
     * Purchase a sweet, reducing its quantity.
     *
     * @param id the sweet ID
     * @param quantity the quantity to purchase
     * @return the updated sweet
     * @throws ResourceNotFoundException if sweet is not found
     * @throws IllegalArgumentException if quantity is insufficient
     */
    public Sweet purchaseSweet(Long id, Integer quantity) {
        Sweet sweet = getSweetById(id);
        sweet.purchase(quantity);
        return sweetRepository.save(sweet);
    }

    /**
     * Restock a sweet, increasing its quantity.
     *
     * @param id the sweet ID
     * @param quantity the quantity to add
     * @return the updated sweet
     * @throws ResourceNotFoundException if sweet is not found
     */
    public Sweet restockSweet(Long id, Integer quantity) {
        Sweet sweet = getSweetById(id);
        sweet.restock(quantity);
        return sweetRepository.save(sweet);
    }

    /**
     * Search for sweets by name, category, or price range.
     *
     * @param name the name to search for (optional)
     * @param category the category (optional)
     * @param minPrice the minimum price (optional)
     * @param maxPrice the maximum price (optional)
     * @return list of matching sweets
     */
    public List<Sweet> searchSweets(String name, String category, BigDecimal minPrice, BigDecimal maxPrice) {
        return sweetRepository.search(name, category, minPrice, maxPrice);
    }
}

File: frontend/src/services/sweetService.js
import apiClient from "./api";

/**
 * Get all sweets
 * @returns {Promise} - List of all sweets
 */
export const getAllSweets = async () => {
  const response = await apiClient.get("/sweets");
  return response.data;
};

/**
 * Get a sweet by ID
 * @param {number} id - Sweet ID
 * @returns {Promise} - Sweet details
 */
export const getSweetById = async (id) => {
  const response = await apiClient.get(`/sweets/${id}`);
  return response.data;
};

/**
 * Search sweets
 * @param {Object} params - Search parameters (name, category, minPrice, maxPrice)
 * @returns {Promise} - List of matching sweets
 */
export const searchSweets = async (params) => {
  const response = await apiClient.get("/sweets/search", { params });
  return response.data;
};

/**
 * Add a new sweet (Admin only)
 * @param {Object} sweetData - Sweet details
 * @returns {Promise} - Created sweet
 */
export const addSweet = async (sweetData) => {
  const response = await apiClient.post("/sweets", sweetData);
  return response.data;
};

/**
 * Update a sweet (Admin only)
 * @param {number} id - Sweet ID
 * @param {Object} sweetData - Updated sweet details
 * @returns {Promise} - Updated sweet
 */
export const updateSweet = async (id, sweetData) => {
  const response = await apiClient.put(`/sweets/${id}`, sweetData);
  return response.data;
};

/**
 * Delete a sweet (Admin only)
 * @param {number} id - Sweet ID
 * @returns {Promise} - Response
 */
export const deleteSweet = async (id) => {
  const response = await apiClient.delete(`/sweets/${id}`);
  return response.data;
};

/**
 * Purchase a sweet
 * @param {number} id - Sweet ID
 * @param {number} quantity - Quantity to purchase
 * @returns {Promise} - Updated sweet
 */
export const purchaseSweet = async (id, quantity) => {
  const response = await apiClient.post(`/sweets/${id}/purchase`, { quantity });
  return response.data;
};

/**
 * Restock a sweet (Admin only)
 * @param {number} id - Sweet ID
 * @param {number} quantity - Quantity to add
 * @returns {Promise} - Updated sweet
 */
export const restockSweet = async (id, quantity) => {
  const response = await apiClient.post(`/sweets/${id}/restock`, { quantity });
  return response.data;
};

export default {
  getAllSweets,
  getSweetById,
  searchSweets,
  addSweet,
  updateSweet,
  deleteSweet,
  purchaseSweet,
  restockSweet,
};
